doc/design/context.md

```diff
@@ -0,0 +1,171 @@
+## Context Design
+
+`Net` is the container and controller of a set of `Operator`. Each `Operator` in `Net` has a method called `Run` to make computation. The `Run` method of `Operator` is defined as follows:
+
+```c++
+Error Operator::Run(OpContext* context);
+```
+
+And the `OpContext` is defined as follows:
+
+```c++
+Struct OpContext {
+  Scope* scope;
+  DeviceContext* device_context;
+};
+```
+
+The `Run` method will take `Variable` (containing `Tensor`) from `Scope` to make computation on certain `DeviceContext`. `DeviceContext` provides necessary runtime resources for computation, including CudaStream, cublasHandle and so on.
```

Comment:
I am writing a simple engine according to mxnet::engine's design document in spare time. 
https://github.com/Superjom/NaiveEngine
It is nearly finished, and I am writing more UT to make sure that it works in large throughput.

Currently, a `DebugEngine` is an engine without thread pool, a `MultiThreadEnginePooled`  with a thread pool of only one worker is the case:

> A simpler engine could be class `SingleThreadEngine` which runs a net on a specific device using one thread.

The code is much less than original mxnet's, and logic simpler

Just add a choice :-) @wangkuiyi @reyoung